### PR TITLE
Replace `orx-concurrent-vec` with `append-only-vec`

### PR DESCRIPTION
**Pull request by MichaReiser**
An alternative to https://github.com/salsa-rs/salsa/pull/554

`append-only-vec` performs even better than `boxcar` and also helps to reduce the overall dependencies (and runs on 32bit!).

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Seems like a winner!

File: Cargo.toml
Comment (nikomatsakis) [inline Cargo.toml:17]:
```suggestion
append-only-vec = "0.1.5"
```

we actually need 0.1.5 since it contains [this change](https://github.com/droundy/append-only-vec/pull/4)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by MichaReiser** [comment]
Thanks for fixing the version. I'll merge this tomorrow unless you wanted me to address something else.

**Comment by MichaReiser** [comment]
Okay I can't merge this while changes are requested. @nikomatsakis would you mind having another look.

**Review by nikomatsakis — APPROVED** [review]

**Comment by ibraheemdev** [comment]
FWIW `append-only-vec` isn't a great concurrent vector implementation as it relies heavily on spin locks. In particular, it holds a [spin lock when allocating](https://github.com/droundy/append-only-vec/blob/main/src/lib.rs#L131) which is almost always a bad idea, but it also uses a spin lock [to serialize insertion order](https://github.com/droundy/append-only-vec/blob/main/src/lib.rs#L158). These not only [hurt scalability](https://github.com/orxfun/orx-concurrent-vec?tab=readme-ov-file#growth-with-push) but can also lead to unfortunate tail latency scenarios that are common with spin locks.

Of course, the upside is that array access is extremely cheap due to the extra ad-hoc synchronization, so if you expect writes to be relatively rare this may not be an issue.. but something to look out for.

**Comment by MichaReiser** [comment]
Thanks @ibraheemdev for the in-depth explanation. 

I do think the spin-lock is fine the way we use it (at least for ingredients) because all write paths go through a lock anyway.

**Comment by ibraheemdev** [comment]
Interesting. A single-writer concurrent vector could be implemented more efficiently in general.. but probably not worth it for Salsa.

**Pull request merged**
